File: dp/stock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock-with-transaction-fee/solutions/108870/Most-consistent-ways-of-dealing-with-the-series-of-stock-problems/
# https://labuladong.github.io/algo/1/13/
#
# 动态规划核心套路 说过
# 动态规划算法本质上就是穷举「状态」，然后在「选择」中选择最优解
#
# 第一步 穷举「状态」，然后在「选择」中选择最优解
# 第二步 有哪些狀態, 第i天 交易k次 是最明顯的狀態
# 第三步 有哪些選擇, sell buy 以及 rest 不做任何事情
# 第四步 列出狀態轉移函數, 發現進行選擇的時候
#       缺乏一個資訊, 讓我們難以進行選擇, 也就是 要手上有股票才能賣, 沒股票才能買
#       這樣一個不容易觀察到的狀態 也就是 目前手上是否有股票 has
# 第五步 把 dp 分為兩個部份進行分析 dp[i][T][0] dp[i][T][1]
# 其 它  只有購買操作會更改允許的最大交易數量
# base  T[-1][k][0] = 0, T[-1][k][1] = -Infinity
#       T[i][0][0] = 0, T[i][0][1] = -Infinity
#
# 此问题共 n × K × 2 种状态，全部穷举就能搞定。
# for 0 <= i < n:
#     for 1 <= k <= K:
#         for s in {0, 1}:
#             dp[i][k][s] = max(buy, sell, rest)

class Solution:

    @staticmethod
    def max_profit_i(prices: list[int]) -> int:
        # leetcode 121
        # https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

        # dp[0][i] 第 i 天 沒股票的最大利潤
        # dp[1][i] 第 i 天 有股票的最大利潤
        n = len(prices)
        dp = [[0 for _ in range(n)] for _ in range(2)]

        # base
        dp[1][0] = float('-inf')  # 第一天 手上有股票 最大利潤為不存在

        for i in range(1, n):
            # 第 i 天 沒股票的最大利潤, 存在兩種情況
            # 1. i-1天 買股票, i 天 賣掉
            # 2. i-1天 沒股票, i 天 沒任何動作
            dp[0][i] = max(dp[1][i - 1] + prices[i], dp[0][i - 1])

            # 第 i 天 有股票的最大利潤, 存在兩種情況
            # 與上述同理
            #
            # 因為只交易一次, 所以不需要 dp[0][i - 1]
            # dp[1][i] = max(dp[0][i - 1] - prices[i], dp[1][i - 1])
            dp[1][i] = max(-prices[i], dp[1][i - 1])

        print(dp)
        return dp[0][n - 1]

    def max_profit_iv(self, k: int, prices: list[int]) -> int:
        # leetcode 188
        # https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iv/
        pass


if __name__ == '__main__':
    obj = Solution()

    sol1 = obj.max_profit_i([7, 1, 5, 3, 6, 4])
    print(f'max_profit_i = {sol1}')

File: dp/Stock.go
package dp

import "math"

// https://github.com/LL-Pengfei/labuladong-algorithm/blob/master/%E5%8A%A8%E6%80%81%E8%A7%84%E5%88%92%E7%B3%BB%E5%88%97/%E5%9B%A2%E7%81%AD%E8%82%A1%E7%A5%A8%E9%97%AE%E9%A2%98.md
// 每天都有三种「选择」：买入、卖出、无操作
//
// dp[i][k][0 or 1]
// 0 <= i <= n-1, 1 <= k <= K
// n 为天数，大 K 为最多交易数
// 此问题共 n × K × 2 种状态，全部穷举就能搞定。
//
// for 0 <= i < n:
//    for 1 <= k <= K:
//        for s in {0, 1}:
//            dp[i][k][s] = max(buy, sell, rest)

func maxProfit_greedy(prices []int) int {
	ans := 0
	min := math.MaxInt
	for i := 0; i < len(prices); i++ {
		if prices[i] < min {
			min = prices[i]
		}

		profit := prices[i] - min
		if profit > ans {
			ans = profit
		}
	}
	return ans
}

func maxProfit(prices []int) int {
	// https://programmercarl.com/0121.%E4%B9%B0%E5%8D%96%E8%82%A1%E7%A5%A8%E7%9A%84%E6%9C%80%E4%BD%B3%E6%97%B6%E6%9C%BA.html#%E6%80%9D%E8%B7%AF
	// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
	var maxFn func(a, b int) int
	maxFn = func(a, b int) int {
		if a < b {
			return b
		}
		return a
	}

	size := len(prices)
	// 「持有状态」: 手上有股票 or 手上沒有股票
	// 選擇: 買 或 賣
	// 持有跟買賣是不同概念

	// dp[i][0] 第i天不持有股票的最大利潤
	// dp[i][1] 第i天 持有股票的最大利潤
	dp := make([][2]int, size) // 1: 有股票, 0: 沒有股票
	dp[0][1] = -prices[0]

	// 如果第i天持有股票即dp[i][1]， 那么可以由两个状态推出来
	// 第i-1天就持有股票，那么就保持现状，所得现金就是昨天持有股票的所得现金 即：dp[i - 1][1]
	// 第i天买入股票，所得现金就是买入今天的股票后所得现金即：-prices[i]

	// 如果第i天不持有股票即dp[i][0]， 也可以由两个状态推出来
	// 第i-1天就不持有股票，那么就保持现状，所得现金就是昨天不持有股票的所得现金 即：dp[i - 1][0]
	// 第i天卖出股票，所得现金就是按照今天股票价格卖出后所得现金即：prices[i] + dp[i - 1][1]

	for i := 1; i < size; i++ {
		dp[i][0] = maxFn(
			dp[i-1][0],           // 前一天就不持有股票, 今天保持原有狀態
			dp[i-1][1]+prices[i], // 前一天持有股票, 今天賣掉
		)
		dp[i][1] = maxFn(
			dp[i-1][1], // 前一天就持有股票, 今天保持原有狀態
			-prices[i], // 前一天不持有股票, 今天購買
			// dp[i-1][0]-prices[i], // 前一天不持有股票, 今天購買, 只能買賣一次, 所以不需要考慮之前
		)
	}
	return dp[size-1][0]
}

func maxProfitII(prices []int) int {
	// 	https://leetcode.com/problems/best-time-to-buy-and-sell-stock-ii/
	// 	https://programmercarl.com/0122.%E4%B9%B0%E5%8D%96%E8%82%A1%E7%A5%A8%E7%9A%84%E6%9C%80%E4%BD%B3%E6%97%B6%E6%9C%BAII%EF%BC%88%E5%8A%A8%E6%80%81%E8%A7%84%E5%88%92%EF%BC%89.html#%E6%80%9D%E8%B7%AF

	// 相關說明參考 maxProfitI

	size := len(prices)
	var maxFn func(a, b int) int
	maxFn = func(a, b int) int {
		if a < b {
			return b
		}
		return a
	}

	dp := make([][2]int, size) // 1: 有股票, 0: 沒有股票
	dp[0][1] = -prices[0]

	for i := 1; i < size; i++ {
		dp[i][0] = maxFn(
			dp[i-1][0],           // 前一天就不持有股票, 今天保持原有狀態
			dp[i-1][1]+prices[i], // 前一天持有股票, 今天賣掉
		)
		dp[i][1] = maxFn(
			dp[i-1][1], // 前一天就持有股票, 今天保持原有狀態
			// -prices[i], // 前一天不持有股票, 今天購買 ( 限制買賣一次的寫法 )
			dp[i-1][0]-prices[i], // 前一天不持有股票, 今天購買 ( 可以買賣多次)
		)
	}
	return dp[size-1][0]
}
